via/services/youtube.py

```diff
@@ -58,8 +63,22 @@ def get_transcript(self, video_id):
         :raise Exception: this method might raise any type of exception that
             YouTubeTranscriptApi raises
         """
-        return YouTubeTranscriptApi.get_transcript(video_id)
+        try:
+            transcript = (
+                self._db.scalars(
+                    select(Transcript).where(Transcript.video_id == video_id)
+                )
+                .one()
+                .transcript
+            )
+        except NoResultFound:
+            transcript = YouTubeTranscriptApi.get_transcript(video_id)
```

Comment:
I think we might want to process the transcript we get from this API before we store it for two reasons. The first is to vet we are getting the shape we expect, and the second is to try and keep the size down.

Looking at the docs we get back something like this:

```python
[
    {
        'text': 'Hey there',
        'start': 7.58,
        'duration': 6.13
    },
    {
        'text': 'how are you',
        'start': 14.08,
        'duration': 7.58
    },
    ...
]
```

There's quite a lot of repetition in the keys there. You can automatically deal with that with something like [ZSON](https://github.com/postgrespro/zson), but I've not used it and I don't think we can put custom plugins in our RDS Postgres instances. In any case it's potentially annoying and fragile.

Much simpler would be to convert this to a more compact tuple format like:

```python
[
    [7.58, 6.13, 'Hey there'],
    [14.08, 7.58, 'how are you'],
    ...
]
```

I think this will save us quite a decent amount of storage space. If we also used this as our API format it would also save us and our users bandwidth when transmitting it.

I looked at whether it might make sense to use something like zlib to compress the string before we send it, but Postgres will store big strings like this in TOAST storage, which is already compressed. It's not great compression, but it's totally transparent to us and already priced in.

Reply:
I tried this out on a random hour long YouTube video and the results of the JSON string length were:

 * Dict format: 136,664
 * Tuple format: 90,235
 
This is about 2/3 of the original size. This is a decent saving, but it's pre-postgres. To really know you'd want to store both in the DB 1000 times and see what happens after Postgres is done messing with it.

In any case, it would mean a bandwidth reduction of 33% for sure if we adopted it as the API format.

Reply:
> I think we might want to process the transcript we get from this API before we store it for two reasons. The first is to vet we are getting the shape we expect, and the second is to try and keep the size down.

I don't think there's much point in the backend validating the transcripts that we get from the YouTube API. All the backend does with these transcripts is blindly pass them to the frontend. If a transcript isn't in the shape that the frontend expects then I think it'd be up to the frontend to detect that and show an error. After all there's not much the backend could do if it detected an invalid transcript: it could just send an error to the frontend which would then have to detect that API error and show an error anyway.

It might have the advantage of potentially preventing us from storing an invalid transcript in the DB? But then again that'd have the downside of disabling our caching and suddenly causing a spike in how many requests we're sending to YouTube if the format of transcripts changes from what our backend expects and so we stop caching them.

Regarding compression and saving on bandwidth: isn't that applied at the HTTP level? Like by NGINX? I guess we could potentially save on Postgres disk space but again that seems like something that Postgres itself might already do.

Reply:
Oh I see, you're actually talking about reformatting the transcripts from dicts into tuples. Hmm. It's more complicated because currently the transcripts are just opaque strings to the backend, and then just the frontend needs to understand YouTube's transcript format. With reformatting the backend would need to understand YouTube's transcript format and then the frontend would need to understand the backend's reformatted transcript format. On the other hand, it sounds like it could save us storage space and bandwidth even though we already have generic compression

Reply:
We [decided](https://hypothes-is.slack.com/archives/C1MA4E9B9/p1690279983515129) to postpone any pre-processing or optimisation of transcripts for maybe later